src/OrchardCore/OrchardCore.DisplayManagement.Liquid/Tags/LayoutTag.cs

```diff
@@ -0,0 +1,35 @@
+using System;
+using System.IO;
+using System.Text.Encodings.Web;
+using System.Threading.Tasks;
+using Fluid;
+using Fluid.Ast;
+using Fluid.Tags;
+
+namespace OrchardCore.DisplayManagement.Liquid.Tags
+{
+    public class LayoutTag : ExpressionTag
+    {
+        public override async Task<Completion> WriteToAsync(TextWriter writer, TextEncoder encoder, TemplateContext context, Expression expression)
+        {
+            var name = (await expression.EvaluateAsync(context)).ToStringValue();
+
+            if (!context.AmbientValues.TryGetValue("LiquidPage", out dynamic page))
+            {
+                throw new ArgumentException("LiquidPage missing while invoking 'layout'");
+            }
+
+            //May be a 'LiquidPage' or a generic 'RazorPage<>'.
+            if (page != null && !string.IsNullOrWhiteSpace(name))
+            {
+                // So we only check if the page has a 'ViewLayout' property.
+                if (((object)page).GetType().GetProperty("ViewLayout") != null)
```

Comment:
It's defined in RazorPage, it should be a a base class then, no generics (if this is the issue) and you can create a Dynamic Delegate that is cached and sets the property on this page instance instead of invoking reflection all the time.

Reply:
@sebastienros 

Here it could be a `LiquidPage : Razor.RazorPage<dynamic>` , so okay., Then when coming e.g from a db template, we don't use a custom file provider => razor engine => bridge to liquid engine. So, in this case it could be our `Razor.RazorPage<dynamic>` or a `Razor.RazorPage<SomeOtherViewModelProxy>` so i was not able to cast to the right type without using reflexion.

Hmm, maybe just because we define in `RazorPage<TModel>` all properties (which don't use `<TModel>`), in place of defining these properties in a non generic base class. I will check this.